master/pkg/searcher/asha.go

```diff
@@ -0,0 +1,244 @@
+) ([]Operation, error) {
+	s.totalStepsCompleted++
+	return nil, nil
+}
+
+func (s *asyncHalvingSearch) validationCompleted(
+	ctx context, requestID RequestID, message Workload, metrics ValidationMetrics,
+) ([]Operation, error) {
+	// Extract the relevant metric as a float.
+	metric, err := metrics.Metric(s.AsyncHalvingConfig.Metric)
+	if err != nil {
+		return nil, err
+	}
+	if !s.AsyncHalvingConfig.SmallerIsBetter {
+		metric *= -1
+	}
+
+	return s.promoteAsync(ctx, requestID, message, metric), nil
+}
+
+func (s *asyncHalvingSearch) promoteAsync(
+	ctx context, requestID RequestID, message Workload, metric float64,
+) []Operation {
+	// Upon a trial is finished, we should return at least one more trial to train unless
+	// the bracket of successive halving is finished.
+	rungIndex := s.trialRungs[requestID]
+	rung := s.rungs[rungIndex]
+	rung.outstandingTrials--
+	addedNewTrial := false
+
+	ops := make([]Operation, 0)
+	// If the trial has completed the top rung's validation, close the trial.
+	if rungIndex == s.AsyncHalvingConfig.NumRungs-1 {
+		s.trialsCompleted++
+		if !s.earlyExitTrials[requestID] {
+			ops = append(ops, NewClose(requestID))
+		}
+	} else {
+		// This is not the top rung, so do promotions to the next rung.
+		nextRung := s.rungs[rungIndex+1]
+		for _, promotionID := range rung.promotionsAsync(
+			requestID,
+			metric,
+			s.AsyncHalvingConfig.Divisor,
+		) {
+			s.trialRungs[promotionID] = rungIndex + 1
+			nextRung.outstandingTrials++
+			if !s.earlyExitTrials[promotionID] {
+				ops = append(
+					ops, trainAndValidate(promotionID, rung.stepsNeeded, nextRung.stepsNeeded)...)
+				addedNewTrial = true
+			} else {
+				step := s.rungs[rungIndex+1].stepsNeeded
+				wkld := Workload{
+					Kind:         ComputeValidationMetrics,
+					ExperimentID: message.ExperimentID,
+					TrialID:      message.TrialID,
+					StepID:       step,
+				}
+
+				// We make a recursive call that will behave the same
+				// as if we'd actually run the promoted job and received
+				// the worse possible result in return.
+				return s.promoteAsync(ctx, promotionID, wkld, ashaExitedMetricValue)
+			}
+		}
+	}
+
+	allTrials := len(s.trialRungs)
+	if !addedNewTrial && allTrials < s.maxTrials {
+		create := NewCreate(
+			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
+		ops = append(ops, create)
+		ops = append(ops, trainAndValidate(create.RequestID, 0, s.rungs[0].stepsNeeded)...)
+		s.rungs[0].outstandingTrials++
+		s.trialRungs[create.RequestID] = 0
+	}
+
+	// Only close out trials once we have reached the maxTrials for the searcher.
+	if allTrials == s.maxTrials {
+		ops = append(ops, s.closeOutRungs()...)
+	}
+	return ops
+}
+
+// closeOutRungs closes all remaining unpromoted trials in any rungs that have no more outstanding
+// trials.
+func (s *asyncHalvingSearch) closeOutRungs() []Operation {
+	ops := make([]Operation, 0)
+	for _, rung := range s.rungs {
+		if rung.outstandingTrials > 0 {
```

Comment:
question: It seems like having the above check `allTrials == s.maxTrials` means that we miss a case where there are no outstandingTrials for a given rung? Why wouldn't we close out that rung if that were the case?

Reply:
The bottom rung isn't complete until there are max_trials on the bottom rung, hence we can't start closing the rung until that point.  The trials on other rungs depend on having seen all trials on the bottom rung so we also can't start closing higher rungs until lower rungs are closed.